tools/base/aio.py

```diff
@@ -83,3 +105,386 @@ async def run_system_command():
         """
         loop = loop or asyncio.get_running_loop()
         return await loop.run_in_executor(executor, partial(subprocess.run, *args, **kwargs))
+
+
+_marker = object()
+
+
+class parallel:  # noqa: N801
+    """This utility provides very similar functionality to
+    `asyncio.as_completed` in that it runs coroutines in parallel, yielding the
+    results as they are available.
+
+    There are a couple of differences:
+
+    - `coros` can be any `iterables` including sync/async `generators`
+    - `limit` can be supplied to specify the maximum number of concurrent tasks
+
+    Setting `limit` to `-1` will make all tasks run in parallel.
+
+    The default `limit` reflects the default for asyncio's
+    `ThreadPoolExecutor`, this is a fairly arbitrary number to use as the tasks
+    are run concurrently on the same thread, but it seems to work quite well.
+
+    For network tasks it might make sense to set the concurrency `limit` lower,
+    if, for example, opening many concurrent connections will trigger
+    rate-limiting or soak bandwidth.
+
+    If an error is raised while trying to iterate the provided coroutines, the
+    error is wrapped in an `AsyncParallelIteratorError` and is raised immediately.
+
+    In this case, no further handling occurs, and `yield_exceptions` has no
+    effect.
+
+    Any errors raised while trying to create or run tasks are wrapped in
+    `AsyncParallelError`.
+
+    Any errors raised during task execution are wrapped in
+    `AsyncParallelExecutionError`.
+
+    If you specify `yield_exceptions` as `True` then the wrapped errors will be
+    yielded in the results.
+
+    If `yield_exceptions` is False (the default), then the wrapped error will be
+    raised immediately.
+
+    If you use any kind of `Generator` or `AsyncGenerator` to produce the
+    awaitables, and `yield_exceptions` is `False`, in the event that an error
+    occurs, it is your responsibility to `close` remaining awaitables that you
+    might have created but which have not already been fired.
+
+    This utility is mostly useful for parallelizing io-bound (as opposed to
```

Comment:
Playing devil's advocate; what's the concrete motivation for adding this feature? Is this about speeding up the future of check_format? What's the best can achieve given GIL? When does parallel execution on many trivially parallel things (e.g. independent files) with something like `multiprocessing` out perform?

It makes a lot of sense to me to have it and use where applicable, but at the same time it would be good to understand the motivation for adding a non-trivial amount of code (and its maintenance implications etc).

Reply:
> what's the concrete motivation for adding this feature? Is this about speeding up the future of check_format? 

its not really about check format - as explained in the ticket description that was just a ~easy place to implement now

mostly its for speeding up **io** tasks

> It makes a lot of sense to me to have it and use where applicable, but at the same time it would be good to understand the motivation for adding a non-trivial amount of code (and its maintenance implications etc).

yep, understood. i have been meaning to implement this for some time as it feels like the missing piece in the asyncio core tools but its quite a bit of code

my initial implementation was much smaller, as i tested it out i found corner cases and the code evolved

its worth stressing that by far the majority of the code being added is tests - the actual code is no more than a few hundred lines

There are broadly 3 situations where we might want to parallelize tasks

- forking subprocesses
- cpu-slow/blocking python code
- io-slow code

The first 2 situations need/want to be pushed to separate threads/cores. These are fairly well taken care of with stdlibs. As they require workers (thread or core), concurrency can be easily limited by just limiting the number of workers.

The third situation is kinda what asyncio code was designed for - parallelizing io blocking code

Despite this there is not a stdlib that does all of

- yields per result
- limits concurrency
- accepts generators

Before working on this implementation i did some research and found some related discussion on the python discussion boards with a reference implementation that was being considered for inclusion in python itself (altho the conversation seemed to stall some time ago)

The problem that i found with that implementation is that it worked in "batch" mode rather than "swarm" mode - ie if you limit concurrency to 4 - 4 jobs are scheduled, 4 jobs complete, another 4 jobs are scheduled

In summary i expect this to be used quite a lot - i already have 3 or 4 places waiting to use it - but my motivation for working on this right now (apart from falling into the rabbit hole) is that on the dist site it will need to download a lot of release files to build the repos, so i was looking for ways to speed that up

Reply:
probs worth flagging my next PR here https://github.com/envoyproxy/envoy/pull/17587/files which makes use of this to upload release files on the Envoy side and downloads them to build repos on the dist side

Reply:
answering these questions more fully...

> What's the best can achieve given GIL?

asyncio - which is single-threaded and runs under the GIL is designed to optimize io tasks which makes that a bit of "piece of string" question

for example - lets say you have 2 (same-sized) network transfer tasks to 2 different endpoints. Supposing you have ~unlimited bandwidth but the endpoint bandwidth is (similarly) limited, then you are going to see ~2x speedup 

OTOH lets suppose that *your* bandwidth is limited and not the endpoints - in that case parallelizing these tasks is not going to give a huge speedup, the bottleneck is your connect. In this case any speedups are going to be either from parallelizing the http connection routines etc or just by getting your requests queued earlier

im guessing the 20% speed up i see is more like the second scenario

> When does parallel execution on many trivially parallel things (e.g. independent files) with something like multiprocessing out perform?

anything cpu-bound - yapf gets a ~5-10% (at best) boost from parallelizing - as its not going to help much with the processing, and they are fighting for the same disk io, any speedup must be coming from queuing the file io more efficiently (or perhaps just from the more JIT style execution that asyncio code has)

Reply:
so regarding `yapf` specifically - the best we can do is to run it "as if" it was a separate process

we can do that pretty easily with asyncio - the code is almost identical to the code already landed for `subprocess.parallel` it just needs to directly run whatever python code in `asyncio.run_in_executor` rather than running `subprocess.run`

this is a good step on that path as it switches the python check class to async, so we can shift it to multithread/core once we have a util in place to do that

Reply:
(or i guess just call yapf externally using `subprocess.parallel`)

Reply:
doing a bit of further research on this...

(qotd has to be "how many threads can a threadpool thread if a threadpool could thread threads" 8/)

afaict python threads always run on the same core (apparently due to GIL) - which means that asyncio is pretty much the theoretical equivalent of threading in python with perhaps some differences in scheduling

the key difference is that with threads the OS does the scheduling

the takeaway is that to improve python perf it on cpu-bound stuff it really has to be using a `ProcessPoolExecutor`

Reply:
Thanks for the explanation. This does make a ton of sense. I'm still trying to grok one thing though; presumably asyncio is like any other I/O async environment, where I/O happens asynchronously by the OS. You only schedule coroutines when data is available and there is corresponding compute work to achieve. Why is providing extra parallelism likely to solve this for very slow I/O?

E.g. we're downloading two large files and blocked mostly on the network, assuming we have easily sufficient CPU to process the data when they arrive on the asyncio runner thread, why would having two of these help?

Reply:
> ... why would having two of these help?

not sure exactly, but something to do with concurrency of waiting

ive been playing with my baseline timing test (which is seeing the 20% speedup) - i added proper async file handling to make sure that wasnt blocking anything, and played around with the transfer chunk size. I also tested out with thread and process pool executors

the 20% speedup is consistent and changed very little. Fixing the file handling reduced the memory footprint to close to zero, and making the chunk size too small, didnt affect the timing, but made my cpu go brrrr

swithcing out the async parallelization to use thread and process pooling was fractionally slower

all of this is consistent with my previous experience of using asyncio (altho the 20% speedup is somewhat arbitrary dep on my current env)

there is quite a good article here https://realpython.com/python-concurrency/#asyncio-version - which goes through all this and does timing comparisons etc

it explains that asyncio is faster than threading or pooling for these kind of tasks because it has less overheads, but all the benefit of concurrency

it also laments the lack of `AsyncioPoolExecutor` which is basically what this pr provides

Reply:
I think I'm pretty skeptical we won't see a huge speedup for pure I/O tasks from concurrency because of the fact that they must work reasonably well on a single threaded executor for I/O. The payoff must come from doing some CPU work in parallel (possibly in conjunction with the I/O), other than possibly being able to make use of kernel parallelism when interacting with the I/O subsystem. But I'm suspecting we never see anything like linear thread scaling based on this.

I'm OK with merging this if it provides a convenient way to structure and think about asyncio in your code. Just I'm not fully convinced that this is a real performance payoff in general beyond some small constant factor.